### PR TITLE
[Finder] Command::addAtIndex() fails with Command instance argument

### PR DESCRIPTION
**Pull request by thunderer**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #14384 |
| License | MIT |
| Doc PR | - |

Fixed reported bug #14384 and added test case for it.

### CODE REVIEW

File: src/Symfony/Component/Finder/Tests/FinderTest.php [hunk 1]
```diff
@@ -852,4 +853,15 @@ public function testMultipleLocationsWithSubDirectories()
         $this->assertIterator($expected, $finder);
         $this->assertIteratorInForeach($expected, $finder);
     }
+
+    public function testShellCommand()
+    {
+        $cmd = Command::create()->add('--force');
+        $this->assertSame('--force', $cmd->join());
```
Comment (xabbuh) [inline src/Symfony/Component/Finder/Tests/FinderTest.php]: This is not related to the issue, is it?

Comment (thunderer) [inline src/Symfony/Component/Finder/Tests/FinderTest.php]: It's just a sanity check before checking problem case. `Command` class is not tested anywhere directly, so I thought it would be a good idea.

Comment (xabbuh) [inline src/Symfony/Component/Finder/Tests/FinderTest.php]: Maybe it's a better idea to split this into more than one test then.

By the way, it could be interesting to use index `1` instead of `2` in the second call of `addAtIndex()` to ensure that you can also insert in the middle.

Comment (thunderer) [inline src/Symfony/Component/Finder/Tests/FinderTest.php]: How about introducing separate `Shell\CommandTest.php` suite to cover `Command` class then? I'd be happy to work on it as it's one of the least covered class in a whole component.

Comment (thunderer) [inline src/Symfony/Component/Finder/Tests/FinderTest.php]: @xabbuh Can you reply to my comment? I'd be happy to refactor the tests, but I need guidance on what effect are you interested in.

Comment (xabbuh) [inline src/Symfony/Component/Finder/Tests/FinderTest.php]: @thunderer A dedicated `CommandTest` class sounds like a good idea to me.

**Comment by thunderer** [comment]
@xabbuh I've added tests for `Shell\Command` class, 100% coverage from now. PHP 5.3.3 build errored due to network problem with Composer. Please tell me what you think.

BTW Since I'm the author of the `Tests\Shell\CommandTest.php` file, can I add @author docblock above `class` declaration?

File: src/Symfony/Component/Finder/Tests/Shell/CommandTest.php [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        $this->assertContains('PHP', $result[0]);
+    }
+
+    public function testCastToString()
+    {
+        $cmd = Command::create();
+        $cmd->add('--force');
+        $cmd->add('--run');
+
+        $this->assertSame('--force --run', (string)$cmd);
```
Comment (dosten) [inline src/Symfony/Component/Finder/Tests/Shell/CommandTest.php]: missing space: `(string) $cmd`

Comment (thunderer) [inline src/Symfony/Component/Finder/Tests/Shell/CommandTest.php]: Thanks, I'll fix this in one batch when @xabbuh gets back to me with his thoughts. :)

**Comment by thunderer** [comment]
@xabbuh Can you comment on the current state of this PR?

**Comment by fabpot** [comment]
:+1: 

Can you rebase and fix CS issues?

File: src/Symfony/Component/Finder/Tests/Shell/CommandTest.php (hunk 2 above)
Comment (xabbuh) [inline src/Symfony/Component/Finder/Tests/Shell/CommandTest.php]: Sorry, I missed this. Yes, please change as suggested. :)

File: src/Symfony/Component/Finder/Tests/FinderTest.php [hunk 3]
```diff
@@ -13,6 +13,7 @@
 
 use Symfony\Component\Finder\Finder;
 use Symfony\Component\Finder\Adapter;
+use Symfony\Component\Finder\Shell\Command;
```
Comment (xabbuh) [inline src/Symfony/Component/Finder/Tests/FinderTest.php]: This should be reverted.

**Comment by jakzal** [comment]
ping @thunderer

**Comment by thunderer** [comment]
@jakzal I fixed CS issues and rebased my branch for 2.3 but it added the whole history to my branch. Can you merge it like that or is it bad and I should do something about that?

**Comment by jakzal** [comment]
@thunderer it might be hard. Are you sure you rebased with the latest 2.3?

**Comment by thunderer** [comment]
I don't know what happened, I just fetched upstream/2.3, pulled it, then rebased my branch and strange things happened. Please see PR #15223, hopefully it'll be okay.
